### PR TITLE
I refactored the code to be able to more easily add additional handlers for URLs and study blocks.

### PR DESCRIPTION
**Pull request by bskeen**
Refactored the structure of the code to allow for future extensions/corner case handling.

### CODE REVIEW

**Comment by kkartch0** [comment]
One initial thought I have is I am wondering if you could split up the changes into multiple commits.
I am seeing some changes that are simply moving methods to another file such as the gospelStudyPluginSettings.ts.

File: src/gospelStudyPluginSettings.ts
Comment (kkartch0) [inline src/gospelStudyPluginSettings.ts:1]: Perhaps this could be renamed to "defaultGospelPluginSettings.ts"

File: src/paragraphParser/paragraphIdDefaultParser.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+/**
+ * A reducing function that inserts a hyphen if the paragraphs aren't contiguous.
+ *
+ * @param {ParagraphResults} prev - The cumulative list of paragraph IDs.
+ * @param {ParagraphItem} curr - The parse results of the current paragraph ID.
+ * @param {number} index - The current index
+ * @returns {ParagraphResults} - The list of paragraph IDs, including hyphens where there is a break between paragraphs.
+ */
+function insertHyphenBetweenNonContiguousParagraphs(prev: ParagraphResults, curr: ParagraphItem, index: number): ParagraphResults {
```
Comment (kkartch0) [inline src/paragraphParser/paragraphIdDefaultParser.ts:93]: What would you say are the benefits of switching to a reducer here instead of using the original for loop?

File: src/studyBlockParser/remoteStudyBlockParser.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        return `${documentTitle}:${displayParagraphs}`;
+    }
+
+    return documentTitle;
+}
+
+/**
+ * Retrieves the document from the given URL, and parses out study block data from there.
+ */
+export const remoteStudyBlockParser: StudyBlockParser = {
```
Comment (kkartch0) [inline src/studyBlockParser/remoteStudyBlockParser.ts:27]: For the sake of less changes all at once, I would like to merge the StudyBlockParsers functionality at a later time. I would like to start with minimal changes to the original code to get the ParagraphIdParsers portion working. Once that is all squared away, we can add the StudyBlockParsers as part of tackling work item #11 on a separate pull request.

**Review by kkartch0 — COMMENTED** [review]
Overall, there are a lot of great ideas here. Though, I am feeling a bit overwhelmed with the number of changes, so I would like these changes to be done in smaller features.

Some examples of what we could do (not necessarily in this order, but it might be a good one):
Feature 1: As a user, I can share a link by manually pasting in OR by using the "share to Obsidian" option. (#26)
   - This could be done by refactoring the "on link pasted" logic into the "onGospelLinkEvent" as you have done. 
   - From there, it would be figuring out how to call that event when "share to Obsidian" option is used.

Feature 2: Proper handling of non "p<number>" ids, such as "kicker", "figure1_p2", "title", etc. (#21)
  - Here we could add some of the basic pasted content parsing functionality. To keep the scope small(ish), we could simply grab the url and strip out everything else that was pasted, treating it as if just the url was pasted. From there, we could parse the ids to formulate the correct study block.

Feature 3: Handling of the "other content" that is included alongside the pasting of the url (#11)
  - Here is where we can expand on the ParagraphIdParsers / PastedContentParsers to handle the other text properly as well.
  - This is where we would also do something with the StudyBlockParserData / PastedContentParserResult

In the end, the combined changes might very well look like what this pull request has, but I think it will have a nicer history that more clearly shows why each change was made.

What are your thoughts on this?

**Comment by kkartch0** [comment]
> One initial thought I have is I am wondering if you could split up the changes into multiple (smaller) commits. I am seeing some changes that seem to simply be moving methods to another file.
> 
> I think smaller commits would help it not seem as much as a massive change.

A couple of examples of how this could be done are on a new branch I created: https://github.com/kkartch0/obsidian-gospel-study/tree/gradual_architecture_change. Each commit passes all of the automated and manual tests I performed.

File: src/models/ParagraphIDParser.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { ParagraphIdResult } from "./ParagraphIdResult";
+
+/**
+ * Declares methods to be used for parsing paragraph IDs out of the given URL.
+ */
+export interface ParagraphIDParser {
```
[the remark this answers is not in the document]
Comment (bskeen) [reply src/models/ParagraphIDParser.ts:6]: Yeah, this is another thing I was planning on cleaning up (along with updating the tests to reflect the new changes).

File: src/models/GospelStudyPluginSettings.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+
+export interface GospelStudyPluginSettings {
```
[the remark this answers is not in the document]
Comment (bskeen) [reply src/models/GospelStudyPluginSettings.ts:2]: Agreed with this one. I think moving the models to their own folder like this is probably the best way to go, just so we have a standard place we can find these. As far as renaming the other file, would you think something like `defaultPluginSettings` would be more reflective of what we have there? Or `pluginSettingConstants`?

File: src/gospelStudyPluginSettings.ts
Comment (bskeen) [reply src/gospelStudyPluginSettings.ts:1]: Haha I should have read more before adding my above comment. I think that this is fine.

File: src/main.ts [hunk 5]
```diff
@@ -94,26 +98,88 @@ export default class GospelStudyPlugin extends Plugin {
...
+
+		let studyBlock: Partial<StudyBlockData> = {};
+
+		// Pass the data through all the applicable parsers to get the results.
+		for (const blockParser of filteredParsers) {
+			studyBlock = await blockParser.mergeStudyBlock(parserData, studyBlock);
+		}
+
+		// Format the study block template with the given data
+		const blockText = this.formatBlockText(studyBlock, this.settings.studyBlockFormat);
```
[the remark this answers is not in the document]
Comment (bskeen) [reply src/main.ts:152]: Thanks! I agree with that. I would even like to maybe move the `formatBlockText` function into a different file.

File: src/paragraphParser/paragraphIdDefaultParser.ts (hunk 1 above)
Comment (bskeen) [reply src/paragraphParser/paragraphIdDefaultParser.ts:93]: It is a little more "JavaScript"-y I guess? I do think in some ways it might be a little more clear what's happening if we switched this to a loop instead, which I am fine doing.

File: src/models/ParagraphIDParser.ts
[the remark this answers is not in the document]
Comment (bskeen) [reply src/models/ParagraphIDParser.ts:1]: I don't really think so. My idea for this is just to try and parse out the paragraph IDs from the URL. Each different constant that implements this would just be trying to get the paragraph IDs that we can use to parse through the HTML from the website. I feel like a name like that would imply that this would be used to handle all of the pasted content (which in the case of something like #11 would be better handled in a study block parser, for example).

[the remark this answers is not in the document]
Comment (bskeen) [reply src/models/ParagraphIDParser.ts:1]: That being said, I'm also not married to this name, we can change it to make it more clear what this should be used for.

File: src/models/ParagraphIdResult.ts
[the remark this answers is not in the document]
Comment (bskeen) [reply src/models/ParagraphIdResult.ts:1]: Maybe see my comments on the `ParagraphIDParser`.

**Comment by bskeen** [comment]
My response to your comments about the size of these changes, along with the path forward, can be summed up in my above comment:

> I can do this if you like. The main point of the changes I made were kind of as a proof-of-concept more than anything. As the code stands in this PR, the current functionality has been preserved, while opening up the architecture to more easily extend it to support the corner cases we have uncovered. I can definitely kind of start over while using this as a reference and create a series of PRs that would add up to doing the same thing I'm doing here (but in a more stable and test-respecting way). We could even create a trail of issues that would add up to this, and then when that is complete, it will make it easier to tackle things such as https://github.com/kkartch0/obsidian-gospel-study/issues/11 or the other corner cases. What do you think of that?

**Comment by bskeen** [comment]
> Overall, there are a lot of great ideas here! Though, I am feeling a bit overwhelmed with the number of changes all at once, so I would like to try and incorporate these changes as part of smaller features.
> 
> Some examples of what we could do (not necessarily in this order, but it might be a good one): Feature 1: As a user, I can share a link by manually pasting in OR by using the "share to Obsidian" option. (#26)
> 
> * This could be done by refactoring the "on link pasted" logic into the "onGospelLinkEvent" as you have done.
> * From there, it would be figuring out how to call that event when "share to Obsidian" option is used.
> 
> Feature 2: Proper handling of non "p<number>" ids, such as "kicker", "figure1_p2", "title", etc. (#21)
> 
> * Here we could add some of the basic pasted content parsing functionality. To keep the scope small(ish), we could simply grab the url and strip out everything else that was pasted, treating it as if just the url was pasted. From there, we could parse the ids to formulate the correct study block.
> 
> Feature 3: Handling of the "other content" that is included alongside the pasting of the url (#11)
> 
> * Here is where we can expand on the ParagraphIdParsers / PastedContentParsers to handle the other text properly as well.
> * This is where we would also do something with the StudyBlockParserData / PastedContentParserResult
> 
> In the end, the combined changes might very well look like what this pull request has, but I think it will have a nicer history that more clearly shows why each change was made.
> 
> What are your thoughts on this?

Sorry, somehow I missed seeing this earlier. I do think this is probably the correct approach to doing this, with a few thoughts:

- In my mind, Feature 2 is the Paragraph ID Parsers. The sole responsibility of this piece (in my mind at least), was retrieving data from the URL. The rest of the pasted/shared content seems like it should be handled as a study block parser. I set it up in the way that I did thinking that we could add parsers to the list to handle the different formats of URLs. Instead of just grabbing one of them like the current code does, we could also handle it a little differently if it seems like that would be better, but we can discuss that later if you like.
- I do agree that the first step could be to separate out the logic so it can be called from a paste or share event.
- The third step does seem to lend itself more to just the study block parsers. I added these in the way that I did with this sort of flow in mind: The parsers could, one by one retrieve the necessary pieces from the available data, and then merge the results with the previous one. I was specifically thinking of something like #11 for this case, but others could be handled in a similar way. As an example, the first parser could get the additional data from the pasted content, namely the selected text and the title/reference, if it is available. Then, the parser that is already in here could see that is available, and use the selected text to only grab what it needs from the HTML (that way, we only include the pasted text, but we can include tags from the HTML we may want to preserve, or include paragraph breaks).

I know I have written a lot of comments back to you about my thought process here. If anything is not clear (or needs to be rethought), please make a note of it. It may be good if we could meet sometime to discuss, so we can share ideas and come up with the best solution we can for this.

File: src/models/ParagraphIDParser.ts
[the remark this answers is not in the document]
Comment (bskeen) [reply src/models/ParagraphIDParser.ts:1]: Maybe a better name to describe what I was thinking for this would be `UrlParser` or `GospelStudyUrlParser`?

File: src/main.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (kkartch0) [reply src/main.ts:152]: That's not a bad idea. That would allow for easier isolated tests around it as well.

File: src/paragraphParser/paragraphIdDefaultParser.ts (hunk 1 above)
Comment (kkartch0) [reply src/paragraphParser/paragraphIdDefaultParser.ts:93]: Haha "JavaScript"-y. Nice. I definitely think the for loop is easier to grasp for those not used to using reducers very often like myself. 😆 I say leave as a loop for now and we can maybe refactor down the road if you still want to. (Leaving as the original loop will also be conducive to paradigm that I would like to follow of "make as little changes as possible" to make the desired feature work).

File: src/studyBlockParser/remoteStudyBlockParser.ts (hunk 2 above)
Comment (kkartch0) [reply src/studyBlockParser/remoteStudyBlockParser.ts:27]: Yes! Let's do that. Having a trail of issues that preserve current functionality but open the architecture up for making it easier to tackle #11 and other corner cases.

File: src/models/ParagraphIDParser.ts
[the remark this answers is not in the document]
Comment (kkartch0) [reply src/models/ParagraphIDParser.ts:1]: Yeah, I like that.  I think GospelStudyUrlParser is a bit wordy.  UrlParser or StudyUrlParser would be good. In my mind, UrlParser would leave it open to potentially expanding beyond the "study" links, to other parts of the church website for example. So, we would take the raw URL that is pasted and run it through the available UrlParsers. But StudyUrlParser to me would indicate you only would pass "/study" urls into the parsers and all "non "churchofjesuschrist.org/study/" links have already been rejected before that point.

[the remark this answers is not in the document]
Comment (kkartch0) [reply src/models/ParagraphIDParser.ts:1]: In the realm of "expanding beyond the "study" links", I was thinking we could maybe utilize the "Copy link to highlight" feature of most browsers as shown below:

image

File: src/models/StudyBlockParserData.ts
[the remark this answers is not in the document]
Comment (kkartch0) [reply src/models/StudyBlockParserData.ts:1]: @bskeen Oops. I think I actually meant combining StudyBlockParserData with ParagraphIdResult _not_ ParagraphIdParser. But your argument probably still stands.

File: src/models/GospelStudyPluginSettings.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kkartch0) [reply src/models/GospelStudyPluginSettings.ts:2]: > Perhaps this could be renamed to "defaultGospelPluginSettings.ts"

I think I like defaultPluginSettings better than defaultGospelPluginSettings that I suggested in a later comment. Let's go with that.

File: src/gospelStudyPluginSettings.ts
Comment (kkartch0) [reply src/gospelStudyPluginSettings.ts:1]: See earlier comment.

**Comment by kkartch0** [comment]
> > Overall, there are a lot of great ideas here! Though, I am feeling a bit overwhelmed with the number of changes all at once, so I would like to try and incorporate these changes as part of smaller features.
> > Some examples of what we could do (not necessarily in this order, but it might be a good one): Feature 1: As a user, I can share a link by manually pasting in OR by using the "share to Obsidian" option. (#26)
> > 
> > * This could be done by refactoring the "on link pasted" logic into the "onGospelLinkEvent" as you have done.
> > * From there, it would be figuring out how to call that event when "share to Obsidian" option is used.
> > 
> > Feature 2: Proper handling of non "p" ids, such as "kicker", "figure1_p2", "title", etc. (#21)
> > 
> > * Here we could add some of the basic pasted content parsing functionality. To keep the scope small(ish), we could simply grab the url and strip out everything else that was pasted, treating it as if just the url was pasted. From there, we could parse the ids to formulate the correct study block.
> > 
> > Feature 3: Handling of the "other content" that is included alongside the pasting of the url (#11)
> > 
> > * Here is where we can expand on the ParagraphIdParsers / PastedContentParsers to handle the other text properly as well.
> > * This is where we would also do something with the StudyBlockParserData / PastedContentParserResult
> > 
> > In the end, the combined changes might very well look like what this pull request has, but I think it will have a nicer history that more clearly shows why each change was made.
> > What are your thoughts on this?
> 
> Sorry, somehow I missed seeing this earlier. I do think this is probably the correct approach to doing this, with a few thoughts:
> 
> * In my mind, Feature 2 is the Paragraph ID Parsers. The sole responsibility of this piece (in my mind at least), was retrieving data from the URL. The rest of the pasted/shared content seems like it should be handled as a study block parser. I set it up in the way that I did thinking that we could add parsers to the list to handle the different formats of URLs. Instead of just grabbing one of them like the current code does, we could also handle it a little differently if it seems like that would be better, but we can discuss that later if you like.
> * I do agree that the first step could be to separate out the logic so it can be called from a paste or share event.
> * The third step does seem to lend itself more to just the study block parsers. I added these in the way that I did with this sort of flow in mind: The parsers could, one by one retrieve the necessary pieces from the available data, and then merge the results with the previous one. I was specifically thinking of something like [Leverage Android's inclusion of selected text #11](https://github.com/kkartch0/obsidian-gospel-study/issues/11) for this case, but others could be handled in a similar way. As an example, the first parser could get the additional data from the pasted content, namely the selected text and the title/reference, if it is available. Then, the parser that is already in here could see that is available, and use the selected text to only grab what it needs from the HTML (that way, we only include the pasted text, but we can include tags from the HTML we may want to preserve, or include paragraph breaks).
> 
> I know I have written a lot of comments back to you about my thought process here. If anything is not clear (or needs to be rethought), please make a note of it. It may be good if we could meet sometime to discuss, so we can share ideas and come up with the best solution we can for this.

No worries! I think the next steps at this point are to come up with a solidified "gameplan" document of how to gradually incorporate the architecture redesign while maintaining current functionality.

**Comment by kkartch0** [comment]
Closing this pull request. The overall result of discussing this pull request draft is that it will be used as a reference to create several smaller pull requests and gradually convert the architecture over to make addressing remaining issues easier.
